akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
         case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
       }
     }
+    val removedKeysIterator = (tombstoneVals.keys ++ that.tombstoneVals.keys).iterator
+    var removedValues = Map.empty[A, B]
+    removedKeysIterator.foreach { key ⇒
+      (this.tombstoneVals.get(key), that.tombstoneVals.get(key)) match {
+        case (Some(thisValue), Some(thatValue)) ⇒
+          if (thisValue.getClass != thatValue.getClass) {
+            val errMsg = s"Wrong type for merging [$key] in [${getClass.getName}], existing type " +
+              s"[${thisValue.getClass.getName}], got [${thatValue.getClass.getName}]"
+            throw new IllegalArgumentException(errMsg)
+          }
+          val removedValue = thisValue.merge(thatValue.asInstanceOf[thisValue.T]).asInstanceOf[B]
+          removedValues = removedValues.updated(key, removedValue)
+        case (Some(thisValue), None) ⇒
+          removedValues = removedValues.updated(key, thisValue)
+        case (None, Some(thatValue)) ⇒
+          removedValues = removedValues.updated(key, thatValue)
+        case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
+      }
+    }
+    // tombstone the values for removed keys
```

Comment:
Unfortunately, tombstones are (as I have implemented them) a global cluster property - the only way to decide to delete the tombstone before applying `update` / `put` message for a given key would be to make sure the `delete` operation was applied on all cluster nodes - because then no operation coming from outside nodes will need tombstone (so wait for confirmation from all currently present cluster members for the reception of `delete` operation (either via delta, full or gossip)). This perhaps looks like a good job for `Replicator`. Let's discuss this avenue... I will also try to look into that (but then probably we need to get your delta confirmation feature merged before this).
Second thing - I will add some tests to validate that correctness assumptions are in no way violated in the scenario you are describing.